File: C++/labs/lab2/patternMatcher.cpp
// Задание 4 | Вариант 3

#include "patternMatcher.h"
#include "array.h" 

#include <iostream>
#include <string>
#include <cstring>
#include <sstream>

bool match(const std::string& text, const std::string& pattern) {
    int i = 0, j = 0;
    int star_idx = -1, match_idx = 0;
    int n = text.length();
    int m = pattern.length();

    while (i < n) {
        if (j < m && (pattern[j] == '?' || pattern[j] == text[i])) {
            i++; j++;
        } else if (j < m && pattern[j] == '*') {
            star_idx = j;
            match_idx = i;
            j++;
        } else if (star_idx != -1) {
            j = star_idx + 1;
            match_idx++;
            i = match_idx;
        } else {
            return false;
        }
    }

    while (j < m && pattern[j] == '*') {
        j++;
    }

    return j == m;
}

/**
 * @brief Основная функция. Загружает строки из файла в DynamicArray и 
 *        проверяет каждую на соответствие шаблону.
 */
void runPatternMatcher(int argc, char* argv[]) {
    std::string query;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--query") == 0 && i + 1 < argc) {
            query = argv[i + 1];
            i++;
            break;
        }
    }

    if (query.empty()) {
        std::cerr << "Ошибка: не найден аргумент --query." << std::endl;
        std::cerr << "Пример: --query \"MATCH <файл_с_данными> <шаблон>\"" << std::endl;
        return;
    }
    
    std::stringstream ss(query);
    std::string command, dataFileName, pattern;
    ss >> command >> dataFileName >> pattern;

    if (command != "MATCH" || dataFileName.empty() || pattern.empty()) {
        std::cerr << "Ошибка: некорректный формат запроса." << std::endl;
        std::cerr << "Пример: --query \"MATCH data.txt h*o\"" << std::endl;
        return;
    }

    // --- ИСПОЛЬЗОВАНИЕ DYNAMICARRAY ---

    // 1. Создаем экземпляр нашего динамического массива
    DynamicArray stringList;

    // 2. Загружаем в него все строки из указанного файла
    stringList.loadFromFile(dataFileName);

    if (stringList.length() == 0) {
        std::cout << "Файл \"" << dataFileName << "\" пуст или не найден." << std::endl;
        return;
    }

    std::cout << "Строки из файла \"" << dataFileName << "\", соответствующие шаблону \"" << pattern << "\":" << std::endl;
    
    bool foundMatches = false;
    // 3. Проходим по всем элементам массива
    for (int i = 0; i < stringList.length(); ++i) {
        std::string currentString = stringList.get(i);
        
        // 4. Вызываем нашу вспомогательную функцию match для каждой строки
        if (match(currentString, pattern)) {
            std::cout << currentString << std::endl;
            foundMatches = true;
        }
    }

    if (!foundMatches) {
        std::cout << "Совпадений не найдено." << std::endl;
    }
}

File: Go/labs/lab2/patternMatcher.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func match(text, pattern string) bool {
	n, m := len(text), len(pattern)
	i, j := 0, 0
	starIdx, matchIdx := -1, 0

	for i < n {
		if j < m && (pattern[j] == '?' || pattern[j] == text[i]) {
			i++
			j++
		} else if j < m && pattern[j] == '*' {
			starIdx = j
			matchIdx = i
			j++
		} else if starIdx != -1 {
			j = starIdx + 1
			matchIdx++
			i = matchIdx
		} else {
			return false
		}
	}

	for j < m && pattern[j] == '*' {
		j++
	}

	return j == m
}

func runPatternMatcher(args []string) {
	var query string
	for i, arg := range args {
		if arg == "--query" && i+1 < len(args) {
			query = args[i+1]
			break
		}
	}

	if query == "" {
		fmt.Fprintln(os.Stderr, "Ошибка: не найден аргумент --query.")
		fmt.Fprintln(os.Stderr, "Пример: --query \"MATCH <файл_с_данными> <шаблон>\"")
		return
	}

	parts := strings.Fields(query)
	if len(parts) != 3 || parts[0] != "MATCH" {
		fmt.Fprintln(os.Stderr, "Ошибка: некорректный формат запроса.")
		fmt.Fprintln(os.Stderr, "Пример: --query \"MATCH data.txt h*o\"")
		return
	}
	dataFileName, pattern := parts[1], parts[2]

	file, err := os.Open(dataFileName)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Файл \"%s\" не найден.\n", dataFileName)
		return
	}
	defer file.Close()

	var stringList []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		stringList = append(stringList, scanner.Text())
	}

	if len(stringList) == 0 {
		fmt.Printf("Файл \"%s\" пуст.\n", dataFileName)
		return
	}

	fmt.Printf("Строки из файла \"%s\", соответствующие шаблону \"%s\":\n", dataFileName, pattern)
	foundMatches := false
	for _, currentString := range stringList {
		if match(currentString, pattern) {
			fmt.Println(currentString)
			foundMatches = true
		}
	}

	if !foundMatches {
		fmt.Println("Совпадений не найдено.")
	}
}